### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

**Comment by justinfagnani** [comment]
@rictic for Typescript stuff

File: src/commands/command.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+export interface Command {
+  name: string;
+  description: string;
+  args: ArgDescriptor[];
+  run(options: {[name: string]: string}): Promise<any>;
+}
+
+// TODO(justinfagnani): how do I re-export an interface?
+export interface ArgDescriptor extends ArgDescriptor {
```
Comment (garlicnation) [inline src/commands/command.ts]: Why do you need to re-export this?

**Comment by samccone** [comment]
oh hey @justinfagnani is there a design doc for this tool? I would love to see the goals here

File: src/commands/command.ts (hunk 1 above)
Comment (justinfagnani) [inline src/commands/command.ts]: hmm... it's not anymore. I was pulling it in with `Command` into polytool and help. I might again.

**Comment by justinfagnani** [comment]
@samccone shared a doc w/ you.

**Comment by samccone** [comment]
a quick note, as much as I like typescript and writing in it for personal projects, if we go down this path we are basically excluding a large number of possible contributors (same argument as coffeescript stuff)

**Comment by justinfagnani** [comment]
This is way different than Coffeescript. Coffeescript is an entirely different syntax, with little tangible benefit, and that's having trouble keeping up with JavaScript changes. Typescript is a relatively small delta from JS, but with a huge benefit from type checking, and polyfilling features like async/await, and they've shown a big willingness to change to be more compatible with JavaScript. Many more projects are moving to Typescript than I have seen move to Coffeescript. We're moving all of our node tools to TS. </:coffee:-:scroll:-rant>

**Comment by samccone** [comment]
sure, just a thought and something we will need to defend because people will ask :)

**Comment by robdodson** [comment]
I can haz doc too?

**Comment by rictic** [comment]
Another big advantage for typescript over similar stuff like coffeescript is that the produced javascript is just the original code without the types. So it's super easy to do JS <-> TS interaction

**Comment by samccone** [comment]
types are good, typescript is good, just like other options, I only wanted to point out that writing in another language for a tool targeting the JS community has tradeoffs, something to be aware of and something to be prepared for. 

+1 for whatever is the most productive and helps us ship something great

**Comment by samccone** [comment]
+1 to landing this :shipit:

**Comment by justinfagnani** [comment]
@garlicnation ping

File: src/polytool.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  commands : Map<String, Command> = new Map();
+  cli : commandLineCommands.CLI;
+
+  constructor() {
+    this.addCommand(new HelpCommand(this.commands));
+    this.addCommand(new ServeCommand());
+  }
+
+  addCommand(command) {
+    this.commands.set(command.name, command);
```
Comment (garlicnation) [inline src/polytool.ts]: What is the intended behavior if a user calls addCommand with the same name twice?

File: src/polytool.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import {Command} from './commands/command';
+
+export class Polytool {
+
+  commandDescriptors = [];
+  commands : Map<String, Command> = new Map();
+  cli : commandLineCommands.CLI;
+
+  constructor() {
+    this.addCommand(new HelpCommand(this.commands));
```
Comment (garlicnation) [inline src/polytool.ts]: Nit: I'd rather see this in bin/, or a static method that provides a polytool configured with a default set of commands. It feel better to me for programmatic consumers to be able to add commands piecemeal.

**Comment by garlicnation** [comment]
LGTM with a couple of nits.

**Comment by addyosmani** [comment]
Can we please share the doc with the rest of the team? Only found out about this repo today.. :(

File: src/polytool.ts (hunk 2 above)
Comment (justinfagnani) [inline src/polytool.ts]: It'll overwrite. It's not a user though, it's us. Let's not do that :)

File: src/polytool.ts (hunk 3 above)
Comment (justinfagnani) [inline src/polytool.ts]: Future work? Not targeting programatic consumers at the moment.

**Pull request merged**